### PR TITLE
feat: integrate anatoliy o11y prototype

### PR DESCRIPTION
**Pull request by letFunny**
This is the `chronoscope.yaml` used:
```
.p1: &tick_parser_positions
  id: 10
  pid: 8
  type: 6
  time: 1
  event: 11

.p2: &rel_parser_positions
  type: 6
  orig_pid: 8
  dest_pid: 10
  orig_id: 12
  dest_id: 14

.p3: &attr_parser_positions
  id: 6
  pid: 4
  type: 2
  name: 7
  value: 1

tick:
  - type: follower
    pos: *tick_parser_positions
  - type: leader
    pos: *tick_parser_positions
  - type: work
    pos: *tick_parser_positions
  - type: rpc
    pos: *tick_parser_positions
  - type: to
    pos: *tick_parser_positions

relation:
  - type: leader-to-rpc
    pos: *rel_parser_positions
  - type: leader-to-work
    pos: *rel_parser_positions
  - type: leader-to-to
    pos: *rel_parser_positions
  - type: follower-to-rpc
    pos: *rel_parser_positions
  - type: follower-to-work
    pos: *rel_parser_positions
  - type: follower-to-to
    pos: *rel_parser_positions

attr:
  - type: gw-attr
    pos: *attr_parser_positions
  - type: req-attr
    pos: *attr_parser_positions
  - type: conn-attr
    pos: *attr_parser_positions
```

Examples: 
follower

leader

### CODE REVIEW

File: src/lib/sm.c [hunk 1]
```diff
@@ -16,18 +17,46 @@ int sm_state(const struct sm *m)
...
+		from->name, to->name, 111UL, 111UL, from->id, to->id);
+}
+
 void sm_init(struct sm *m,
 	     bool (*invariant)(const struct sm *, int),
 	     bool (*is_locked)(const struct sm *),
 	     const struct sm_conf *conf,
 	     int state)
 {
+	static sig_atomic_t id = 0;
```
Comment (just-now) [inline src/lib/sm.c:50]: It's better to use uint64_t here. 
sig_atomic_t is uni32_t on some machines and can easily overflow.

File: src/lib/sm.c [hunk 2]
```diff
@@ -16,18 +17,46 @@ int sm_state(const struct sm *m)
...
+
 	PRE(conf[state].flags & SM_INITIAL);
 
 	m->conf = conf;
 	m->state = state;
 	m->invariant = invariant;
 	m->is_locked = is_locked;
+	m->id = (long unsigned int)++id;
```
Comment (just-now) [inline src/lib/sm.c:58]: comment above will affect this line.

File: src/lib/sm.h [hunk 3]
```diff
@@ -29,6 +29,8 @@ struct sm
 {
 	int rc;
 	int state;
+	const char *name;
+	unsigned long int id;
```
Comment (just-now) [inline src/lib/sm.h:33]: i don't remember if it was introduced by me, still please use uint64_t here.

File: src/lib/sm.c [hunk 4]
```diff
@@ -16,18 +17,46 @@ int sm_state(const struct sm *m)
...
+void sm_to_sm_obs(const struct sm *from, const struct sm *to)
+{
+	struct timespec ts = {0};
+	struct tm tm;
+
+	clock_gettime(CLOCK_REALTIME, &ts);
+	gmtime_r(&ts.tv_sec, &tm);
+
+	tracef("%s-to-%s opid: %lu dpid: %lu id: %lu id: %lu |\n",
+		from->name, to->name, 111UL, 111UL, from->id, to->id);
```
Comment (just-now) [inline src/lib/sm.c:41]: s/`111UL`/$real_pid/.

File: src/lib/sm.c [hunk 5]
```diff
@@ -16,18 +17,46 @@ int sm_state(const struct sm *m)
...
+static inline void sm_obs(const struct sm *m)
+{
+	struct timespec ts = {0};
+	struct tm tm;
+
+	clock_gettime(CLOCK_REALTIME, &ts);
+	gmtime_r(&ts.tv_sec, &tm);
+
+	tracef("%s pid: %lu sm_id: %lu %s |\n",
+		m->name, 111UL, m->id, m->conf[sm_state(m)].name);
```
Comment (just-now) [inline src/lib/sm.c:29]: s/111UL/$real_pid/.

File: src/raft/recv_install_snapshot.c [hunk 6]
```diff
@@ -541,19 +541,19 @@ static void work_fill_follower(struct follower *follower)
...
 	rpc_init(&leader->rpc);
-	// sm_to_sm_obs(&leader->sm, &leader->rpc.sm);
+	sm_to_sm_obs(&leader->sm, &leader->rpc.sm);
 }
 
 static void rpc_fill_follower(struct follower *follower)
 {
 	rpc_init(&follower->rpc);
-	// sm_to_sm_obs(&follower->sm, &follower->rpc.sm);
+	sm_to_sm_obs(&follower->sm, &follower->rpc.sm);
```
Comment (just-now) [inline src/raft/recv_install_snapshot.c:556]: For future: let's think about better function name for `sm_to_sm_obs()`.

**Review by just-now — APPROVED** [review]
Looks good.

File: src/lib/sm.c [hunk 7]
```diff
@@ -42,7 +71,7 @@ void sm_move(struct sm *m, int next_state)
 {
 	int prev = sm_state(m);
 
-	tracef("SM_MOVE %s => %s", m->conf[prev].name, m->conf[next_state].name);
+	sm_obs(m);
```
[the remark this answers is not in the document]
Comment (letFunny) [reply src/lib/sm.c:74]: You are right, I will move it to after the state transition, thanks!

File: src/lib/sm.c (hunk 1 above)
Comment (letFunny) [reply src/lib/sm.c]: Changed.

File: src/lib/sm.h (hunk 3 above)
Comment (letFunny) [reply src/lib/sm.h]: Changed.

File: src/lib/sm.c (hunk 5 above)
Comment (letFunny) [reply src/lib/sm.c]: Changed.

File: src/raft/recv_install_snapshot.c (hunk 6 above)
Comment (letFunny) [reply src/raft/recv_install_snapshot.c]: Changed to `sm_relate`, I am open to better suggestions :)

File: src/raft/recv_install_snapshot.c [hunk 8]
```diff
@@ -415,8 +415,8 @@ static void follower_work_done(struct work *w)
 
 static void to_init(struct timeout *to)
 {
-	// static const char *to_sm_name = "to";
-	// to->sm = (struct sm){ .name = to_sm_name };
+	static const char *to_sm_name = "to";
+	to->sm = (struct sm){ .name = to_sm_name };
 	sm_init(&to->sm, to_sm_invariant, NULL, to_sm_conf, TO_INIT);
```
[the remark this answers is not in the document]
Comment (letFunny) [reply src/raft/recv_install_snapshot.c]: Done, thanks!

**Review by cole-miller — APPROVED** [review]

**Pull request merged**